subsys/debug/cpu_load/cpu_load.c

```diff
@@ -0,0 +1,297 @@
+static void cpu_load_log_fn(struct k_work *item)
+{
+	u32_t load = cpu_load_get();
+	u32_t percent = load / 1000;
+	u32_t fraction = 1000 % percent;
+
+	cpu_load_reset();
+	LOG_INF("Load:%d,%03d%%", percent, fraction);
+	k_delayed_work_submit(&cpu_load_log, CPU_LOAD_LOG_INTERVAL);
+}
+
+static int cpu_load_log_init(void)
+{
+	k_delayed_work_init(&cpu_load_log, cpu_load_log_fn);
+	return k_delayed_work_submit(&cpu_load_log, CPU_LOAD_LOG_INTERVAL);
+}
+
+static void timer_handler(nrf_timer_event_t event_type, void *context)
+{
+	/*empty*/
+}
+
+
+int cpu_load_init(void)
+{
+	u8_t ch_sleep;
+	u8_t ch_wakeup;
+	u8_t ch_tick = 0;
+	nrfx_err_t err;
+	nrfx_timer_config_t config = NRFX_TIMER_DEFAULT_CONFIG;
+	int ret = 0;
+
+	if (ready) {
+		return 0;
+	}
+
+	config.frequency = NRF_TIMER_FREQ_1MHz;
+	config.bit_width = NRF_TIMER_BIT_WIDTH_32;
+
+	if (IS_ENABLED(CONFIG_CPU_LOAD_ALIGNED_CLOCKS)) {
+		/* It's assumed that RTC1 is driving system clock. */
+		config.mode = NRF_TIMER_MODE_COUNTER;
+		err = ppi_alloc(&ch_tick,
+		       nrf_rtc_event_address_get(NRF_RTC1, NRF_RTC_EVENT_TICK));
+		if (err != NRFX_SUCCESS) {
+			return -ENODEV;
+		}
+		nrfx_gppi_channel_endpoints_setup(ch_tick,
+		     nrf_rtc_event_address_get(NRF_RTC1, NRF_RTC_EVENT_TICK),
+		     nrfx_timer_task_address_get(&timer, NRF_TIMER_TASK_COUNT));
+		nrf_rtc_event_enable(NRF_RTC1, NRF_RTC_INT_TICK_MASK);
+	}
+
+
+	err = ppi_alloc(&ch_sleep,
+		       nrf_power_event_address_get(NRF_POWER,
+						   NRF_POWER_EVENT_SLEEPENTER));
+	if (err != NRFX_SUCCESS) {
+		ppi_cleanup(ch_tick, CH_INVALID);
+		return -ENODEV;
+	}
+
+	err = ppi_alloc(&ch_wakeup,
+		       nrf_power_event_address_get(NRF_POWER,
+						   NRF_POWER_EVENT_SLEEPEXIT));
+	if (err != NRFX_SUCCESS) {
+		ppi_cleanup(ch_tick, ch_sleep);
+		return -ENODEV;
+	}
+
+	nrfx_gppi_channel_endpoints_setup(ch_sleep,
+		  nrf_power_event_address_get(NRF_POWER,
+					      NRF_POWER_EVENT_SLEEPENTER),
+		  nrfx_timer_task_address_get(&timer, NRF_TIMER_TASK_START));
+	nrfx_gppi_channel_endpoints_setup(ch_wakeup,
+		  nrf_power_event_address_get(NRF_POWER,
+					      NRF_POWER_EVENT_SLEEPEXIT),
+		  nrfx_timer_task_address_get(&timer, NRF_TIMER_TASK_STOP));
+
+
+	err = nrfx_timer_init(&timer, &config, timer_handler);
+	if (err != NRFX_SUCCESS) {
+		return -EBUSY;
+	}
+	/* In case of DPPI event can only be assigned to a single channel. In
+	 * that case, cpu load can still subscribe to the channel but should
+	 * not control it. It may result in cpu load not working. User must
+	 * take care of that.
+	 */
+	nrfx_gppi_channels_enable((IS_CH_SHARED(ch_sleep) ? 0 : BIT(ch_sleep)) |
+				(IS_CH_SHARED(ch_wakeup) ? 0 : BIT(ch_wakeup)) |
+				(IS_CH_SHARED(ch_tick) ? 0 : BIT(ch_tick)));
+
+	cpu_load_reset();
+
+	if (IS_ENABLED(CONFIG_CPU_LOAD_LOG_PERIODIC)) {
+		ret = cpu_load_log_init();
+	}
+
+	ready = true;
+
+	return ret;
+}
+
+void cpu_load_reset(void)
+{
+	int key = irq_lock();
+
+	nrfx_timer_enable(&timer);
+	nrfx_timer_disable(&timer);
```

Comment:
I guess it should be the other way round, so that the timer stays enabled afterwards.

Reply:
i missed your edit message but came to the same conclusion and replaced it with clearing. Also removed the lock because timer will not progress until CPU sleeps.